src/queue/descriptor_chain.rs

```diff
@@ -178,3 +184,815 @@ impl<M: GuestAddressSpace> Iterator for DescriptorChainRwIter<M> {
         }
     }
 }
+
+#[derive(Clone, Debug)]
+struct DescriptorChainConsumer<'a> {
+    buffers: Vec<VolatileSlice<'a>>,
+    bytes_consumed: usize,
+    index: usize,
+}
+
+impl<'a> DescriptorChainConsumer<'a> {
+    fn available_bytes(&self) -> usize {
+        // This is guaranteed not to overflow because the total length of the chain
+        // is checked during all creations of `DescriptorChainConsumer` (see
+        // `Reader::new()` and `Writer::new()`).
+        self.buffers[self.index..]
+            .iter()
+            .fold(0usize, |count, vs| count + vs.len() as usize)
+    }
+
+    fn bytes_consumed(&self) -> usize {
+        self.bytes_consumed
+    }
+
+    // Given a DescriptorChain offset, return the index in self.buffers, and the
+    // offset within, or ChainOverflow error.
+    fn position(&self, offset: usize) -> Option<(usize, usize)> {
+        let mut rem = offset;
+        for (pos, vs) in self.buffers[self.index..].iter().enumerate() {
+            if rem <= vs.len() {
```

Comment:
Looks like we should replace `<=` with a `<` here.

Reply:
No, if rem fits in vs.len() we should return that. (if you replace with < it will fail)